sihl/src/web_bearer_token.ml

```diff
@@ -2,28 +2,48 @@ let key : string Opium.Context.key =
   Opium.Context.Key.create ("token", Sexplib.Std.(sexp_of_string))
 ;;
 
-let find req = Opium.Context.find_exn key req.Opium.Request.env
-let find_opt req = Opium.Context.find key req.Opium.Request.env
+exception Token_not_found
+
+let find req =
+  try Opium.Context.find_exn key req.Opium.Request.env with
+  | _ ->
+    Logs.err (fun m -> m "No bearer token found");
+    Logs.info (fun m ->
+        m "Have you applied the bearer token middleware for this route?");
+    raise @@ Token_not_found
+;;
+
+let find_opt req =
+  try Some (find req) with
+  | _ -> None
+;;
 
 let set token req =
   let env = req.Opium.Request.env in
   let env = Opium.Context.add key token env in
   { req with env }
 ;;
 
-let middleware =
+let default_handler _ =
+  Opium.Response.of_plain_text "Unauthorized 401"
+  |> Opium.Response.set_status `Unauthorized
+  |> Lwt.return
+;;
+
+let middleware ?(unauthenticated_handler = default_handler) () =
   let filter handler req =
     match Opium.Request.header "authorization" req with
     | Some authorization ->
-      (match String.split_on_char ' ' authorization with
-      | [ "Bearer"; token ] ->
-        let req = set token req in
-        handler req
-      | [ "bearer"; token ] ->
-        let req = set token req in
-        handler req
-      | _ -> handler req)
-    | None -> handler req
+      let bearer_token =
+        try Some (Scanf.sscanf authorization "Bearer %s" (fun b -> b)) with
+        | _ -> None
+      in
+      (match bearer_token with
+      | None -> unauthenticated_handler req
```

Comment:
I propose you skip this indirection via `option` and add `unauthenticated_handler req` to the failure case of the `Scanf.sscanf` directly and ignore the `Some` for the success case too.

Reply:
This won't work since `handler req` might raise and we might catch it.

Reply:
Got it